rasterio/_io.pyx

```diff
@@ -546,19 +558,30 @@ cdef class DatasetReaderBase(DatasetBase):
         `out` (if used), or will be masked if any of the nodatavals are
         not `None`.
         """
-        cdef int height, width, xoff, yoff, aix, bidx, indexes_count
+        cdef int aix, bidx, indexes_count
+        cdef double height, width, xoff, yoff
         cdef int retval = 0
         cdef GDALDatasetH dataset = NULL
 
+        if out is None:
+            raise ValueError("An output array is required.")
+
         dataset = self.handle()
 
-        # Prepare the IO window.
         if window:
-            window = windows.evaluate(window, self.height, self.width)
-            yoff = <int>window[0][0]
-            xoff = <int>window[1][0]
-            height = <int>window[0][1] - yoff
-            width = <int>window[1][1] - xoff
+            if not isinstance(window, Window):
+                raise WindowError("window must be an instance of Window")
```

Comment:
Is it possible to specify the type of the `window` arg using Cython? just wondering, not a blocker